### PR TITLE
Annotated input type

### PR DESCRIPTION
**Pull request by devmaslov**
The initial goal of this PR is to solve the issue about handling update without default values and with only properties passed by a user into the request. You can find my initial proposal here: https://github.com/thecodingmachine/graphqlite/issues/210. 

This update brings the following features and changes:

- New annotation `@Input` is introduced on classes. It works the same way as `@Type`, but for input types. In combination with `@Field` annotation on properties you can describe input object type. When this input type is injected into a mutation during the request new instance of the class is created with only properties passed by a user. See examples below. The `@input` annotation has the following attributes:
  - `name`: Input name. If it's omitted it will be generated based on the class name. For example for class name `User`, `UserInput` type name will be generated.
  - `default`: Determines whether this type is default. The same idea as in `@Factory` annotation.
  - `description`: Just a description for documentation.
  - `update`: Determines if this input should be in "update" mode. For such input types all properties are optional and without default values.
- `@Field` annotation now can be used on class properties https://github.com/thecodingmachine/graphqlite/issues/246. If property is private or protected setter (for input) or getter (for output) is required. The type is guessed from property type (>= php 7.4) or from PHPDoc `@var` tag. Also description is extracted from PHPDoc and applied to the field. This annotation has a couple of new attributes:
  - `for`: `@Field` annotation can be now applied several times to the same method or property. Property `for` allows to determine to which type it should belong to. For example: `@Field(for="User")` or `@Field(for={"CreateUserInput", "UpdateUserInput"})`
  - `description`: If you require a different description per different type for the same field, you can now use a new property `description` of `@Field` annotation.
  - `inputType`: allows to set a specific input type on a field - similar as it works with `outputType`.
- Field middlewares such as `@Security`, `@Right`, `@Logged` etc. can be used on properties alongside with `@Field`.

In the initial proposal there was an idea about pre-loading an entity on update, so it would be possible to inject into mutation a loaded from DB User for example with populated data from the request. I tried to follow the idea where you put `ID` (or another field) as first param and `payload` as the second one. For example: `updateUser(id: ID!, user: UserInput!)`, `updateUserByEmail(email: String!, user: UserInput)` or `updateMe(user: UserInput)` (for logged it user). Implementing this might be an overhead, because at least we need to dynamically add a property (if necessary) to a mutation based on which entity should be loaded and also loaders which would handle the pre-loading. I think this kind of thing can be done as a separate feature I'll gladly discuss any ideas about it. For now I suggest using this feature for describing DTOs as input types and inject them into mutations. Mutation can use a custom service or handle DTO to Entity transformation by themself.

Here's an example how it all works together.

`UserInput` - DTO for user entity with public fields. Totally the same can be achieved for private/protected fields with setters. Also you can omit it and describe inputs directly on `User` class.
```
/**
 * @GQL\Input(name="CreateUserInput")
 * @GQL\Input(name="UpdateUserInput", update=true)
 */
final class UserInput
{

    /**
     * @var string Username or login
     *
     * @GQL\Field()
     */
    public string $username;

    /**
     * @var string User email
     *
     * @GQL\Field(for="CreateUserInput") <-- Required for create and cannot be updated
     */
    public string $email;

    /**
     * @var string User's password
     *
     * @GQL\Field(name="password") <-- In DTO you would probably just name this field "password", but I just want to show here that the name of the input field also can be different from the property name. 
     */
    public string $plainPassword;

    /**
     * @var DateTimeInterface|null Birthday
     *
     * @GQL\Field(inputType="Date") <-- For example we have a custom scalar type Date without time.
     */
    public ?DateTimeInterface $birthday; <-- This field is nullable, so it's optional even for CreateUserInput.
}
```

`User` - actual entity.
```
/**
 * User entity.
 *
 * @GQL\Type()
 */
class User
{
    /**
     * @var string Username or login
     *
     * @GQL\Field()
     */
    private string $username;

    /**
     * @var string User email
     *
     * @GQL\Field()
     */
    private string $email;

    /**
     * @var DateTimeInterface|null Birthday
     *
     * @GQL\Field(outputType="Date")
     */
    private ?DateTimeInterface $birthday = null;
}
```
And mutations. Instance of `UserInput` will be created with populated data from a request and injected for "$user" argument. With `@Assertion` annotation those inputs can be validated in Symfony.
```
    /**
     * @GQL\Mutation()
     * @GQL\UseInputType(for="$user", inputType="CreateUserInput!")
     *
     * @Assertion(for="$user", constraint=@Assert\Valid)
     *
     * @param UserInput $user
     *
     * @return User
     */
    public function createUser(UserInput $user): User
    {
        // Create a new entity from a UserInput object, save and return it
    }

    /**
     * @GQL\Mutation()
     * @GQL\UseInputType(for="$user", inputType="UpdateUserInput!")
     *
     * @Assertion(for="$user", constraint=@Assert\Valid)
     *
     * @param string    $id
     * @param UserInput $user
     *
     * @return User
     */
    public function updateUser(string $id, UserInput $user): User
    {
        // Load user by id and populate it with data in UserInput.
        // You can create some custom mechanism of tracking what properties were set to the $user
        // or check which properties are initialized (PHP >= 7.4)
        // or just loop through $user - only initialized properties will be iterated.
        // In such case if user just wants to update one field, email for example - only it will be updated, without default values on other ones.
    }
```

I must say that the code is not covered with unit tests yet. First I'd like to hear some opinions about the implementation and the concept in general.

Also there'are few not critical but still **Breaking changes**. 
`$refMethod` argument has been replaced without a type hint so it would be possible to pass either `ReflectionMethod` or `ReflectionProperty` in such public methods:
- `AnnotationReader::getMiddlewareAnnotations`
- `InvalidPrefetchMethodRuntimeException::methodNotFound`
- `CannotMapTypeException::createForMissingPhpDoc`
- `CachedDocBlockFactory::getDocBlock`
- `RootTypeMapperInterface` methods and all implementing it classes - this is my biggest concern

### CODE REVIEW

**Comment by moufmouf** [comment]
Woah, this is a huge amount of work! Thanks so much!

I'm a bit busy right now, but I'll definitely try to do an extensive code review by the end of the week.

**Comment by gulien** [comment]
This looks amazing 👍 

I wonder if we could add a "partial update" mechanism in the mix (see https://github.com/graphql-dotnet/conventions/issues/128#issuecomment-462804984).

**Comment by devmaslov** [comment]
@moufmouf, thanks, looking forward to it.

@gulien the partial update is possible with this feature. The idea is that input object (such as `UserInput` DTO) is populated only with data that user passes to the graphQL request. Let's say I'm sending this query:

```
mutation {
  updateUser(id: "user-id", user: {
    username: "NewName"
  }) {
    id
    username
  }
}
```

When it's executed only `username` will be set to the instance of `UserInput`. It means that you can implement a way of tracking  which properties has been set and if it's a DTO, properly transform it to a needed entity. For instance the easiest way if you're using PHP 7.4 and your DTO has all properties public and typed (as in my example above) - simply loop through the object with `foreach` and set needed properties into a new object. Here's an example for update user mutation:

```
    /**
     * @GQL\Mutation()
     * @GQL\UseInputType(for="$user", inputType="UpdateUserInput!")
     *
     * @Assertion(for="$user", constraint=@Assert\Valid)
     *
     * @param string    $id
     * @param UserInput $user
     *
     * @return User
     */
    public function updateUser(string $id, UserInput $user): User
    {
        $entity = // some logic to load existing user by $id
        foreach ($user as $key => $value) {
            // Symfony has property accessor service which allows to set values into objects. Let's use that for example:
            $this->propertyAccessor->setValue($entity, $key, $value)
        }
       
        // Save and return $entity
    }
```

It won't work though if you property doesn't have a type. In this case such property has `null` as a default value by PHP. I'm sure there're plenty of other ways to achieve this as well.

**Comment by devmaslov** [comment]
@moufmouf 
Thanks a lot, I am happy to see you feel that way :smiley:
About the review - no worries, take your time and thank you for the first feedback. I'll respond shortly :blush:

> My gut feeling is that it is ok to introduce those breaking changes. Implementing RootTypeMapperInterface if probably not that common.

I also think that it's be fine. For the ones who currently have this this interface implemented, it should be easy to spot that something is wrong, since PHP will throw the exception at compilation time.

File: src/Annotations/Field.php
Comment (devmaslov) [reply src/Annotations/Field.php:36]: Usually PHPDoc comments work out well for API documentation, but I believe it's not always like this. I think the "description" field can provide more flexibility. Here's some use cases:

- As you mentioned there might be a situation where you'd like to have different description per different input/type. You might have different validation rules or security restrictions and it would be a great idea mentioning such things in the API. Let's say I have "email" field in user class with 3 different types:
  - `CreateUserInput` - used on sign up. It's required for a user to specify their email
  - `UpdateUserInput` - only admin user can update "email" field
  - `User` - only admin can see emails or users can see only their emails (we don't want to expose emails to other users for example)

  I forgot to mention it above, but I've adjusted `@Fields` so it would be possible to add several of them for the same property. Like this:
```
/**
 * @GQL\Input(name="CreateUserInput")
 * @GQL\Input(name="UpdateUserInput", update=true)
 * @GQL\Type()
 */
class User {

    /**
     * @var string User email
     *
     * @GQL\Field(for="CreateUserInput", description="User email.")
     * @GQL\Field(for="UpdateUserInput", description="User email. Only admin can update it.")
     * @GQL\Field(for="User", description="User email. Can be accessed by an admin or currently logged in user to see only their emails.")
     */
    public string $email;
}
```
- If you apply `@Field` annotation on a method, the comment might not be appropriate. PHPDoc comments for methods usually start with a verb ( like "Return", "Get" ). I don't think field description should start with a verb, because they do not perform any actions. So instead of "Get user email", the correct description should be "User email". `description` attribute can help fixing that. 
- Sometimes you might want to leave some technical comment in PHPDoc, which should be visible only for internal developers and hidden from API documentation. In `description` attribute you can specify only documentation needed for the API.

Does this make sense to you?

P. S. I realized that I didn't adjust this `description` field to work with `@Type`, it works only with `@Input`. If it makes sense, I'll fix that.

File: src/Annotations/Field.php
Comment (devmaslov) [reply src/Annotations/Field.php:41]: I thought about this and came to a decision that it would be more convenient to describe fields for both type and input within one annotation. You can apply `@Input` and `@Type` annotation for the same class and in most cases you'd want to mark the same properties as fields.

With two annotations you'd need to do this twice which is kind of redundant:
```
/**
 * User entity.
 *
 * @GQL\Type()
 * @GQL\Input()
 */
class User
{
    /**
     * @var string Username or login
     *
     * @GQL\Field()
     * @GQL\InputField()
     */
    private string $username;

    /**
     * @var string User email
     *
     * @GQL\Field()
     * @GQL\InputField()
     */
    private string $email;

    /**
     * @var DateTimeInterface|null Birthday
     *
     * @GQL\Field()
     * @GQL\InputField()
     */
    private ?DateTimeInterface $birthday = null;
}
```

With one `@Field` annotation it looks more simple and clean:
```
/**
 * User entity.
 *
 * @GQL\Type()
 * @GQL\Input()
 */
class User
{
    /**
     * @var string Username or login
     *
     * @GQL\Field()
     */
    private string $username;

    /**
     * @var string User email
     *
     * @GQL\Field()
     */
    private string $email;

    /**
     * @var DateTimeInterface|null Birthday
     *
     * @GQL\Field()
     */
    private ?DateTimeInterface $birthday = null;
}
```

You need to separate input fields from output you can use `for` attribute.

But here it is the problem with `inputType` and `outputType` attributes. If I want to set the same type for both cases and the same field - which is 99% use case I think, I'd need to do this: `@GQL\Field(inputType="Date", outputType="Date")`. 

I considered alternative option - deprecate `outputType` and add another attribute `type` - which would suit for both input and output types. `@GQL\Field(type="Date")` looks much better. What do you think about that?

File: src/Annotations/Input.php
Comment (devmaslov) [reply src/Annotations/Input.php]: Fixed the typo, thanks

File: src/QueryFieldDescriptor.php
Comment (devmaslov) [reply src/QueryFieldDescriptor.php]: Fixed that, thanks!

File: src/Types/InputType.php
Comment (devmaslov) [reply src/Types/InputType.php:123]: I'm not sure if we should overload constructor with such logic. It's job is only to instantiate an object with a given parameters. Current implementation assumes that required constructor parameters have defined fields on properties with the same name in the class. When object is instantiated users payload is mapped to the constructor parameters.

For example:
```
/**
 * @GQL\Input()
 */
class User
{
    /**
     * @var string Username or login
     *
     * @GQL\Field()
     */
    private string $username;

    /**
     * @param string $username
     */
    public function __construct(string $username)
    {
        $this->username = $username;
    }
}
```

In this case we definitely know that input type has a field called `username` and it's required. There's a constructor parameter with the same name so when the object is instantiated value from the request payload will be passed to it. For this `createInstance()` function is responsible - this is kind of default factory, which knows how to map data to a constructor and instantiate an object.

I agree that it would be nice to provide a custom and flexible way of creating instances before they are populated with new data. At the beginning I thought of reusing `@Factory`. We could add a new attribute `@Input(init="FactoryName")`. But after that I realized that it's not very flexible. For example it would be nice to reuse the same input type in different mutations, but pre-load should be done with different identificators. For example: 
```
updateUser(id: ID!, user: UserInput!)
updateUserByEmail(email: String!, user: UserInput)
updateMe(user: UserInput)
```

Considering this we'd need to create 3 different input types with the same fields but different `init` attributes.

As an alternative this `init` attribute could be added to `@UseInputType` annotation. This way we choose what type should be used and how the object should be initialized for each particular mutation. Also extra parameters such as "id: ID!" or "email: String!" described in factories could be mapped not to input types, but to mutations directly. I hope it won't be confusing. Here's an example of what I mean:

```
/**
 * @GQL\Type()
 * @GQL\Input()
 */
class User
{

    /**
     * @var string
     *
     * @GQL\Field(for="User")
     */
    private string $id;
    
    /**
     * @var string
     *
     * @GQL\Field()
     */
    private string $username;

    ...
}

...

class UserFactory
{

    /**
     * @Factory(name="LoadUser")
     * 
     * @param string $id
     * 
     * @return User
     */
    public function loadUser(string $id): User {
       // ... Load user from DB
    }
}

...

class UserController
{

    /**
     * @Mutation(name="UpdateUser")
     * @UseInputType(for="$user", inputType="UserInput!", init="LoadUser")
     * 
     * @param User $user
     * 
     * @return User
     */
    public function updateUser(User $user): User {
       // ... $user is loaded from DB and populated with given data. Just validate and save it. 
    }
}
```

Because `@UseInputType` points to "LoadUser" factory with `init` attribute, another field `id` will be added to the mutation as `loadUser()` function requires it as a parameter. So the request mapping will be like this:
```
updateUserByEmail(id: String!, user: UserInput)
```

My only concern is that it might be not obvious that by adding `init` attribute your mutation signature is split across different methods - actual mutation method and supportive "init" factory. It might be a bit confusing or not obvious...
What do you think about this idea? Do you see other ways to achieve this?

File: src/Annotations/Field.php
Comment (moufmouf) [reply src/Annotations/Field.php:36]: Ok, your explanation makes a lot of sense to me. We will just have to make sure we correctly document it :+1:

File: src/Annotations/Field.php
Comment (moufmouf) [reply src/Annotations/Field.php:41]: Ok, I understand your will to not repeat `Field` and `InputField` annotations. It makes sense.

I'm trying to assess if there are any corner-cases that might bother us in the future.

Let's imagine this:

```
/**
 * User entity.
 *
 * @GQL\Type()
 * @GQL\Input()
 */
class User
{
    // ...

    /**
     * @var string Username or login
     *
     * @GQL\Field()
     */
    private string $password;

    public function setPassword(string $password): void {
    	$this->password = $password;
    }
}
```

In this case, we have a @Field annotation on the property, a setter, but no getter.

What should happen in your opinion? Should GraphQLite throw an error? (something like "Output field "password" is not reachable via a getter, please add a "getPassword" method). Or should we silently put only an input field and no output field?

(I'm trying to put as much attention as possible to all the error messages a user can have and to have meaningful error messages)

File: src/Types/InputType.php
Comment (moufmouf) [reply src/Types/InputType.php:123]: About the "init" idea: I would even go one step further. The concept of @factory is hard to explain and I'm getting a lot of people confused with it.

Since quite often, a factory is used by only one mutation, we could even imagine something like this:

```php
class UserController
{

    /**
     * @Mutation(name="UpdateUser")
     * @UseInputType(for="$user", factory="UserFactory::loadUser", inputName="UpdateUserInput")
     * 
     * @param User $user
     * 
     * @return User
     */
    public function updateUser(User $user): User {
       // ... $user is loaded from DB and populated with given data. Just validate and save it. 
    }
}
```

and this might actually be easier to understand for many people.

File: src/Annotations/Field.php
Comment (devmaslov) [reply src/Annotations/Field.php:41]: It's a good question.

> Or should we silently put only an input field and no output field?
I don't think this is the way to go. At some point it might confuse developers, especially with models containing a lot of properties, why their field is not working even though there's a `@Field` annotation.

I like the idea of throwing an error. At the moment if you try accessing such field PHP will throw an error like `Cannot access private property ...`. It would be much better to display a message as you suggested to give more accurate explanation of the problem.

File: src/Types/InputType.php
Comment (devmaslov) [reply src/Types/InputType.php:123]: Is `UserFactory` just some service with `loadUser` function and has nothing to do with `@Factory` annotation?

I considered similar approach, but in this case you'd need to specify an absolute class name. For example: `@UseInputType(for="$user", factory="App\Services\UserFactory::loadUser", inputName="UpdateUserInput")`. It will be easier to understand and use. Not sure only if it looks good in the code though :smile:

**Comment by moufmouf** [comment]
@devmaslov just a quick message to let you know I started applying some changes in your code (right now, it's only applying cs-fixes to have a green CI) Be sure to pull the code from your branch before working on it again!

**Comment by moufmouf** [comment]
@devmaslov I started adding tests here: https://github.com/thecodingmachine/graphqlite/pull/269/commits/bb69621e6d7e9851e5cf6b82b94e25df5b54a136

It is just a simple test to see if @Field annotation works correctly in properties.
The result is quite suprising. The new test works, but adding the @Field annotation on properties seems to have a side-effect on inheritance. In my test, I'm adding fields on the `Contact` class and suddenly, I can't access fields on the `User` class (that extends `Contact`) anymore.

See: 

https://travis-ci.com/github/thecodingmachine/graphqlite/jobs/353951488#L361

**Comment by devmaslov** [comment]
@moufmouf Looks like this issue with inheritance appears when you property is defined as `private` in parent class. For `public` and `protected` it should work fine. I'll take a look into it. And thanks for adding unit tests, I'll make sure they pass after the fix.

**Comment by devmaslov** [comment]
@moufmouf Fixed the problem. The reason for it was different, but still I noticed interesting behavior with private properties.
Consider the following example:
```
/**
 * @Type()
 */
class Contact {
    /**
     * @Field()
     * @var int
     */
    private $age = 42;

    /**
     * @return int
     */
    public function getAge(): int
    {
        return $this->age;
    }
}

/**
 * @Type()
 */
class User extends Contact {
}
```

In this case field name appears in User type without any problems. But if you remove `@Type` annotation from Contact it will disappear from User - because there's no GraphQL type to inherit it from and `$refClass->getProperties()` doesn't return parent's private properties. I think it should be fixed and field should appear without parent being `@Type` - they still can be available via getters.

File: src/Annotations/Field.php
Comment (devmaslov) [reply src/Annotations/Field.php:34]: Fixed Field description to work with `@Type`

**Comment by oojacoboo** [comment]
@devmaslov just a friendly ping to let you know that this looks awesome and can't wait to get it merged in :)

**Comment by oojacoboo** [comment]
@devmaslov @moufmouf any chance we can get this over the finish line soon before it becomes too stale?

**Comment by devmaslov** [comment]
I'd like to move this forward as well. @moufmouf, what should be our next steps here? Is there anything should be adjusted before we dive into covering all of this with unit tests? There're a couple open threads above :arrow_up:

**Comment by moufmouf** [comment]
Hey @devmaslov ,

Thanks for keeping working on this!

Definitely, we need to add more unit tests here.
The unit tests should cover corner cases. Like:
"what happens if we have both a public property and a getter (with different types....)"

Or

"what happens if we have a private property with a given type and a getter with a different type?"...

These corner cases will help make this PR rock solid! :+1:

**Comment by oojacoboo** [comment]
@devmaslov happy new year!  2021 resolution - get this pull request merged in?  So much good stuff - hate to see it go to waste.

**Comment by devmaslov** [comment]
Hey guys, Happy New Year :tada: @moufmouf, @oojacoboo

I just wanted to make sure I'm on the right path here, before proceeding.
Also want to see it's released :slightly_smiling_face:. So I'll work on unit tests and adjusting documentation within next few days and hopefully soon we'll be able to merge this.

**Comment by moufmouf** [comment]
Hey @devmaslov,

Those last 4 commits look excellent!
:+1: This is definitely going in the right direction.

Let me know when you want me to do a code review.

**Comment by devmaslov** [comment]
@moufmouf 

I think I covered all needed use cases in unit tests and updated documentation for the new features.

Specifically to the use cases you mentioned before:
> The unit tests should cover corner cases. Like:
> "what happens if we have both a public property and a getter (with different types....)"
> Or
> "what happens if we have a private property with a given type and a getter with a different type?"...

I covered cases when there's a public property and a getter/setter. I don't think we actually need to check if getter returns a different type - it is absolutely the same case if you add annotation like `@Field(outputType: "Date")` to a function, but return a string. In this case **webonyx** throws an error that expected one type, but returned another. So in our case getters do not have affect on the field output - they are just resolving functions for properties.

Is it OK that the checks are failing? Looks like it's not related to my code, but to some issues with dependencies.

I want to go a bit further with this feature:
Currently "input" classes are instantiated only by creating a new instance. For update operations it will make a lot of sense to instantiate the class with data from the database before instead of creating a new class instance. For example we could load a doctrine entity first and then resolve the input fields user passed to a query. This is where `update` argument comes in play on it's full potential - it'll make all fields optional and without default values. Entity can be loaded from data based and only the properties user specified in schema will be updated.

As I can see we could use Parameter Annotations for this in a manner very similar to [Param Converters](https://symfony.com/doc/current/bundles/SensioFrameworkExtraBundle/annotations/converters.htmll) in Symfony. Consider the following example:

```
#[Mutation]
public function updateUser(string $id, #[UseInputType('UpdateUserInput!'), LoadEntity] User $user): User
{
...
}
```

Here "LoadEntity" is that parameter annotation that handles loading a Doctrine entity instead of just creating a new User instance by default. Then only fields user passes will be updated in the entity. All is needed to do in the controller is persist and flush the entity (maybe run a validation as well). Parameter Annotations can use all arguments passed to the controller (if I'm not mistaken), so in the example above we can use `$id` argument for loading an entity.

We need to provide a convenient interface for building this kind of Parameter Annotations and build a couple that will be most frequently used - for example Doctrine ORM entities. Any other examples? 

Any thoughts, suggestions?

**Comment by oojacoboo** [comment]
@devmaslov this sounds awesome btw, nice work so far.  I'm really liking where you're headed with this.  

We create a base Factory method that ends up doing what you're proposing, although more verbose, since it's still a factory method and not an annotation on the mutation (I like your proposal).  In each factory method, we're basically doing this:

```php
use InputType\Thing;

public function makeThing(ID $id, Foo $foo): Thing
{
    return $this->make(new Thing(), \get_defined_vars());
}
```

The `make` method hydrates the input type object, then validates it and returns any errors.  The validation was a critical piece here since we're using Symfony's annotation based validation lib.  This allows for us to ensure that we're getting back an entirely valid input type with very little boilerplate.  Granted, we still have to create the input type objects.

I think what you're proposing here is great though.  It'd basically allow for us to just eliminate the Factory class entirely.  We do have some factory methods that require a little parameter massaging before hydrating an input type, but 95% of them are nothing more than the example above.

**Comment by devmaslov** [comment]
@oojacoboo, thanks for your feedback!

The only downside of my proposal I see is that the additional arguments  like `string $id` are technically useless inside of the controller function. So basically they'll be used only for mapping.

To fix this and simplify the usage of the feature we could allow Parameter Annotations to dynamically inject some of the necessary arguments. For example by default `LoadEntity` annotation might add "id" arg to a query. For more flexibility we could provide an arg for fields based on which entity should be loaded like this: `LoadEntity(fields: 'email')` or `LoadEntity(fields: ['name', 'age'])`. 

The example above can be rewritten like this:

```php
#[Mutation]
public function updateUser(#[UseInputType('UpdateUserInput!'), LoadEntity] User $user): User
{
...
}
```

Since `LoadEntity` adds additional argument `$id` the schema would look like this:
```
updateUser(id: ID, user: UpdateUserInput!): User!`
```

Not sure how hard it's going to be to implement this yet.

Regarding the validation it still should be possible to use `Assertion` annotation. If needed you could even create a wrapper for such cases which by default has a `Valid` constraint - convenient for validating objects. Or validation can be called directly inside of the controller.

**Comment by oojacoboo** [comment]
@devmaslov how do you feel about spinning this off into another branch?  That'd allow us to get this current pull request out the door and then spend some time working through this with a fresh look.

@moufmouf Thoughts here and regarding the failing tests @devmaslov mentioned?

**Comment by devmaslov** [comment]
@oojacoboo, agree. This pull already has enough of helpful features :slightly_smiling_face:  No need to delay the release.

**Comment by moufmouf** [comment]
Hey @devmaslov ,

Thanks for the awesome work!

Agree with both of you. We must merge this PR first and stop adding more to it, it is already huge :)
I'll try to see what is going wrong with the CI. By the way, having the "Test the bundles' dependencies" fail is ok. But the "Continuous Integration" steps need to pass. I'll take a look.

**Comment by moufmouf** [comment]
@devmaslov I have fixed the "core dumped" issue by changing the coverage driver from PCOV to XDebug.

Could you simply make sure the CI is passing for "Continuous Integration / Continuous Integration (8.0, highest) (pull_request)"?
It is a matter of applying `composer cs-fix` and manually fixing a dozen of errors returned by `composer cs-check`.

I'll take care of the "Test the bundles' dependencies" errors.

**Comment by devmaslov** [comment]
@moufmouf it's done. CI is passing :tada:

**Comment by devmaslov** [comment]
Hey @moufmouf,
Just a kind reminder to review this PR :slightly_smiling_face:

**Comment by moufmouf** [comment]
Hey @devmaslov ,

I'm definitely not forgetting this PR, I've just been overwhelmed lately, but I'll try to have a proper look at it and validate it this week!

**Review by moufmouf — APPROVED** [review]
Hey @devmaslov ,

The amount of work you put into this PR is astounding. The doc is perfect, unit tests are also great. I have nothing to add. Let's merge this! :+1:

Also, I realize I have been very long to give feedback. I've had an awful lot of work lately (not directly related to GraphQLite alas) and I certainly need some help maintaining this library. Since this PR clearly shows you understood GraphQLite, would you like to be promoted as a maintainer of GraphQLite?

**Comment by oojacoboo** [comment]
@devmaslov Say yes :)

**Comment by moufmouf** [comment]
@oojacoboo the offer also extends to you :heart:
You have been giving many feedbacks to the community in general and if you want to be granted maintainer rights, just tell me :+1:

**Comment by devmaslov** [comment]
@moufmouf, @oojacoboo Thanks a lot for your input and feedback :smile: Happy to see this branch merged :partying_face: 
I'll be happy to help maintaining the library :smiley:

**Comment by moufmouf** [comment]
Cool!
@devmaslov Do you happen to have a Twitter account? (so I can mention you in the upcoming Tweet?)

**Comment by devmaslov** [comment]
@moufmouf no, I don't have one :slightly_smiling_face:

**Comment by oojacoboo** [comment]
@moufmouf I'm happy to continue to help out - discuss issue tickets, help with some pull requests, etc.  I'm fairly familiar with some of the internals.  We do use this lib pretty heavily and will be using it much more so going forward.

**Comment by moufmouf** [comment]
@oojacoboo I'll take that for a yes :D

**Comment by TamasSzigeti** [comment]
Sorry to intrude here, really glad that the responsibility is now shared, thanking each of you. May I call attention to the graphqlite-bundle as well, I have a tiny humble PR waiting there (:

**Comment by Lappihuan** [comment]
@devmaslov this looks awesome! may i ask what is holding back the release of this?

**Comment by oojacoboo** [comment]
@Lappihuan we're mostly waiting for release instructions from @moufmouf

**Comment by MattBred** [comment]
Any movement on 4.2 / instructions @moufmouf ?

**Pull request merged**
